src/gluonts/model/san/_engine.py

```diff
@@ -0,0 +1,302 @@
+from typing import Optional, List, Callable, Iterator
+import os
+import logging
+import tempfile
+import inspect
+import uuid
+import time
+
+import numpy as np
+import mxnet as mx
+from mxnet.gluon import nn, HybridBlock, Trainer as MXTrainer
+
+from gluonts.mx.trainer import Trainer as BaseTrainer
+from gluonts.mx.trainer.learning_rate_scheduler import MetricAttentiveScheduler
+from gluonts.mx.trainer.model_averaging import (
+    AveragingStrategy,
+    SelectNBestMean,
+    save_epoch_info,
+)
+from gluonts.model.forecast import QuantileForecast
+from gluonts.model.forecast_generator import (
+    QuantileForecastGenerator as BaseForecastGenerator,
+)
+from gluonts.dataset.loader import (
+    DataLoader,
+    TrainDataLoader,
+    ValidationDataLoader,
+    InferenceDataLoader,
+)
+from gluonts.dataset.field_names import FieldName
+from gluonts.dataset.common import DataEntry
+from gluonts.gluonts_tqdm import tqdm
+from gluonts.support.util import HybridContext
+
+logger = logging.getLogger("gluonts").getChild("trainer")
+MODEL_ARTIFACT_FILE_NAME = "model"
+STATE_ARTIFACT_FILE_NAME = "state"
+LOG_CACHE = set([])
+OutputTransform = Callable[[DataEntry, np.ndarray], np.ndarray]
+
+
+class Trainer(BaseTrainer):
+    def __call__(
+        self,
+        net: nn.HybridBlock,
+        input_names: List[str],
+        train_iter: TrainDataLoader,
+        validation_iter: Optional[ValidationDataLoader] = None,
+    ) -> None:  # TODO: we may want to return some training information here
+        is_validation_available = validation_iter is not None
+        self.halt = False
+
+        with tempfile.TemporaryDirectory(
+            prefix="gluonts-trainer-temp-"
+        ) as temp:
+
+            def base_path() -> str:
+                return os.path.join(
+                    temp,
+                    "{}_{}".format(STATE_ARTIFACT_FILE_NAME, uuid.uuid4()),
+                )
+
+            def loss_value(loss: mx.metric.Loss) -> float:
+                return loss.get_name_value()[0][1]
+
+            logger.info("Start model training")
+
+            net.initialize(ctx=self.ctx, init=self.init)
+
+            with HybridContext(
+                net=net,
+                hybridize=self.hybridize,
+                static_alloc=True,
+                static_shape=True,
+            ):
+                batch_size = train_iter.batch_size
+
+                best_epoch_info = {
+                    "params_path": f"{base_path()}-init.params",
+                    "epoch_no": -1,
+                    "score": np.Inf,
+                }
+
+                lr_scheduler = MetricAttentiveScheduler(
+                    objective="min",
+                    patience=self.patience,
+                    decay_factor=self.learning_rate_decay_factor,
+                    min_lr=self.minimum_learning_rate,
+                )
+
+                optimizer = mx.optimizer.Adam(
```

Comment:
A lot of this we already have built into the gluonTS `Trainer` class

Reply:
Yes.. The only thing I modified is the arg list fed to the train_net, anything else remains unchanged.